embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/logs/LogOrchestrator.kt

```diff
@@ -0,0 +1,78 @@
+package io.embrace.android.embracesdk.internal.logs
+
+import androidx.annotation.VisibleForTesting
+import io.embrace.android.embracesdk.comms.delivery.DeliveryService
+import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.worker.ScheduledWorker
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+
+internal class LogOrchestrator(
+    private val logOrchestratorScheduledWorker: ScheduledWorker,
+    private val clock: Clock,
+    private val sink: LogSinkImpl,
+    private val deliveryService: DeliveryService
+) {
+    private var lastLogTime: Long = 0
+    private var inactivityFuture: ScheduledFuture<*>? = null
+    private var batchTimeFuture: ScheduledFuture<*>? = null
+
+    init {
+        sink.callOnLogsStored(::onLogsAdded)
+    }
+    private fun onLogsAdded() {
+        lastLogTime = clock.now()
+        if (sink.completedLogs().size >= MAX_LOGS_PER_BATCH) {
+            sendLogs()
+        } else {
+            scheduleInactivityCheck()
+            if (batchTimeFuture == null) {
+                scheduleBatchTimeCheck()
+            }
+        }
+    }
+
+    @VisibleForTesting
+    fun sendLogs() {
+        inactivityFuture?.cancel(false)
+        inactivityFuture = null
+        batchTimeFuture?.cancel(false)
+        batchTimeFuture = null
+
+        // Sink is synchronized, so even if sendLogs is called concurrently, the logs
```

Comment:
That being true, would we get into a race condition where the following happens?

- A log is added that triggers a flush
- A log is added right after that
- One of the timing job runs and flushes the new log that was added because it wasn't cleared in time.

Given that `sendLogs` isn't synchronized, I think we can get into a situation where that happens?

Reply:
While the flushings themselves will be synchronized, so each flush will not over lap, we may be prematurely calling the second flush - when it checks whether it should flush, it'll say yes because the first flush hasn't completed yet. And if there's a log added after the first flush is finished but before the second flush runs, that log will be sent prematurely.

To put it another way, if the second flush results in an empty list, it shouldn't have been flushed at all - the first flush should've canceled the second flush, but because sendLogs isn't synchronized, we can get into this situation

While the second log will only be

Reply:
I've merged the conditions checking and the log sending in a single synchronized method so in the case discussed it would return early instead of flushing the sink.